### DISCUSSION TITLE
Scanning SBOM of unsupported OS produces empty vuln json

### DISCUSSION THREAD

**Discussion by Bah-bah** [aquasecurity/trivy#6630] [Q&A] [answered]
Hi!
I was experimenting with scanning of various OS images and came across this case. I got Fedora official docker image 
`docker pull fedora`
made SBOM of it
`docker run -v /dir/app aquasec/trivy rootfs app/fedora-fs/ --format cyclonedx --scanners vuln -q > image.cyclonedx`
and then scan resulting SBOM
`docker run -v /dir:/app aquasec/trivy sbom app/image.cyclonedx --format json -q > image-rootfs.json`
As I can see [here](https://aquasecurity.github.io/trivy/v0.51/docs/coverage/os/) Trivy doesn't support Fedora. And I expected it at least to show me warning message about unsupported OS or even fail on some step (on scan, for example). Instead, I got valid SBOM with Fedora's packages and short scan results' json without `Results` section and any mention of unsupported OS.
Here it is:
```
{
  "SchemaVersion": 2,
  "CreatedAt": "2024-05-04T08:03:28.276474256Z",
  "ArtifactName": "app/fedora.cyclonedx",
  "ArtifactType": "cyclonedx",
  "Metadata": {
    "OS": {
      "Family": "fedora",
      "Name": "40"
    },
    "ImageConfig": {
      "architecture": "",
      "created": "0001-01-01T00:00:00Z",
      "os": "",
      "rootfs": {
        "type": "",
        "diff_ids": null
      },
      "config": {}
    }
  }
}
```
It's interesting that running `image` command shows the desired warning:
```
viktorlapin$ docker run aquasec/trivy image fedora
5-04T08:26:56Z	INFO	Vulnerability scanning is enabled
2024-05-04T08:26:56Z	INFO	Secret scanning is enabled
2024-05-04T08:26:56Z	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2024-05-04T08:26:56Z	INFO	Please see also https://aquasecurity.github.io/trivy/v0.51/docs/scanner/secret/#recommendation for faster secret detection
2024-05-04T08:27:07Z	INFO	[python] License acquired from METADATA classifiers may be subject to additional terms	name="libcomps" version="0.1.20"
2024-05-04T08:27:07Z	INFO	Detected OS	family="fedora" version="40"
2024-05-04T08:27:07Z	WARN	Unsupported os	family="fedora"
2024-05-04T08:27:07Z	INFO	Number of language-specific files	num=0
```
**The main question here is**: Do I have a way to distinguish between scanning of unsupported OS and scan result without vulnerabilities?

**Comment by Bah-bah** [comment]
Any thoughts on this matter?

**Reply by Bah-bah** [reply]
@DmitriyLewen can you please clarify?

**Reply by DmitriyLewen** [reply]
Hello @Bah-bah 
Sorry for waiting.

What version of Trivy are you using?
I see warning about unsupported OS:
```bash
➜  trivy -v 
Version: 0.52.2
➜  trivy -d image -f cyclonedx -o report.cdx.json fedora
...
➜  trivy sbom report.cdx.json 
2024-06-19T10:01:56+06:00	INFO	Vulnerability scanning is enabled
2024-06-19T10:01:56+06:00	INFO	Detected SBOM format	format="cyclonedx-json"
2024-06-19T10:01:56+06:00	INFO	Detected OS	family="fedora" version="40"
2024-06-19T10:01:56+06:00	WARN	Unsupported os	family="fedora"
2024-06-19T10:01:56+06:00	INFO	Number of language-specific files	num=0
```

**Reply by Bah-bah** [reply]
Hello @DmitriyLewen 
Thank you for your answer.
I'm sorry for this confusion - question was poorly formulated. There are two parts of trivy output - log and result. I can mute log with `-q` flag. The question here: is it possible to understand that OS is not supported without log having only result output json?

**Reply by DmitriyLewen** [reply, marked as answer]
Unfortunately, `json` output does not contain information that an unsupported OS was scanned.

**Reply by WsyKer** [reply]
> 不幸的是，`json`输出不包括支持信息。

Bro, I have also encountered this situation. Is there any solution? Thank you！
![image](https://github.com/aquasecurity/trivy/assets/75110695/8a1eccf9-1b13-4483-841d-46c3ed0d471d)

**Reply by WsyKer** [reply]
> > 不幸的是，`json`输出不包括支持信息。
> 
> 兄弟，我也遇到这种情况了，有没有什么解决办法，谢谢！ ![图像](https://private-user-images.githubusercontent.com/75110695/341253774-8a1eccf9-1b13-4483-841d-46c3ed0d471d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4NjU0NjcsIm5iZiI6MTcxODg2NTE2NywicGF0aCI6Ii83NTExMDY5NS8zNDEyNTM3NzQtOGExZWNjZjktMWIxMy00NDgzLTg0MWQtNDZjM2VkMGQ0NzFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIwVDA2MzI0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0YjVhYWIzNTQ2YmI5ZjgxZTc1MWM5MmYwMWUxM2MzMzJhMGYxNjlmZjI1OWU5N2FjNmRjNjViNDg4Y2U2NGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.l8FX94bGmkQnJctEXaf0mFd4Y2a0S4lKteTDU9SVnao)

I think the problem stems from the changes made to ghcr.io/aquasecurity/trivy db

**Reply by DmitriyLewen** [reply]
Hello @WsyKer 
Trivy can't detect your OS.

Can you send your image?

**Reply by WsyKer** [reply]
> 你好@WsyKer Trivy 无法检测到您的操作系统。
> 
> 您可以发送您的图片吗？

The picture is as follows. Thank you very much
![image](https://github.com/aquasecurity/trivy/assets/75110695/f90ad3d3-f452-4cbd-a908-700d1e30f7d7)

**Reply by WsyKer** [reply]
> > 你好@WsyKer Trivy 无法检测到您的操作系统。
> > 您可以发送您的图片吗？
> 
> The picture is as follows. Thank you very much ![image](https://private-user-images.githubusercontent.com/75110695/341298435-f90ad3d3-f452-4cbd-a908-700d1e30f7d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4NjYxNzQsIm5iZiI6MTcxODg2NTg3NCwicGF0aCI6Ii83NTExMDY5NS8zNDEyOTg0MzUtZjkwYWQzZDMtZjQ1Mi00Y2JkLWE5MDgtNzAwZDFlMzBmN2Q3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIwVDA2NDQzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlhNmIyN2RhMjA1OWM5ODQ0M2JmY2FmYzNhMDIzN2MzMzRhY2IwMzliMWNlNDlkMzg4M2I3OTVmY2YzMDAyNGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jR-ywr7iLObehSBWIiZXN1yagsFhVXhKEdosSiPS0u0)

My command is as follows: trivy image -- timeout 30m - s CRITICAL, HIGH abc:123. This operation prompts an unsupported os and does not return any results 😩😭

**Reply by DmitriyLewen** [reply]
I need to see inside of `abc:123` image.
Can you send this image or send a Dockerfile to build this image?

**Reply by WsyKer** [reply]
> `abc:123`我需要查看图像内部。 您可以发送此图像或发送 Dockerfile 来构建此图像吗？

Very Sorry, abc: 123 has not been published on the Internet. The problem is that I can scan abc: 123 normally using trivy at the beginning of last month, but it will not be available in June

**Reply by DmitriyLewen** [reply]
Unfortunately, without an image, I can’t tell you why Trivy doesn’t find the OS version

**Comment by DmitriyLewen** [comment]
I created #6971 to see the demand for this.

**Reply by Bah-bah** [reply]
It will be great to see this flag in the json.
But in the meantime can I rely on existence/nonexistence of "Results" section in json? I noticed that when OS is unsupported the output json doesn't have this section. And when OS is supported but there're no vulnerabilities found this section exists but has no entries.

**Reply by DmitriyLewen** [reply]
if your image contains language-specific files, `Results` will not be nil and will include packages/vulnerabilities for those files.
Also you can't rely on this logic - because if the image doesn't contain the OS package, the result will also be empty.

That's why I suggested adding a new field

**Reply by Bah-bah** [reply]
> if your image contains language-specific files, Results will not be nil and will include packages/vulnerabilities for those files.

I never realised it. So, even though OS isn't supported by Trivy, it still scans language packages?

> if the image doesn't contain the OS package, the result will also be empty

You mean like in busybox? Results section doesn't appear at all in two cases: 1. OS not supported and there's no lang packages 2. There's no OS packages at all ?

**Reply by DmitriyLewen** [reply]
> I never realised it. So, even though OS isn't supported by Trivy, it still scans language packages?

Trivy scans language-specific files separate of OS package:
```Dockefile
FROM fedora

COPY ./log4j-core-2.18.0.jar /log4j-core-2.18.0.jar    
```
```bash
➜ trivy -d image -f json --list-all-pkgs  6630              
...
{
  "SchemaVersion": 2,
  "CreatedAt": "2024-06-20T16:32:47.464814692+06:00",
  "ArtifactName": "6630",
  "ArtifactType": "container_image",
  "Metadata": {
    "OS": {
      "Family": "fedora",
      "Name": "40"
    },
    ...
  "Results": [
    {
      "Target": "Java",
      "Class": "lang-pkgs",
      "Type": "jar",
      "Packages": [
        {
          "Name": "org.apache.logging.log4j:log4j-core",
          "Identifier": {
            "PURL": "pkg:maven/org.apache.logging.log4j/log4j-core@2.18.0",
            "UID": "7574f76ebd40de4e"
          },
          "Version": "2.18.0",
          "Layer": {
            "DiffID": "sha256:ba1011d30192002b6848c3d8cd2ddc6e31da54cbf9dbf81e0d112b33e92c99aa"
          },
          "FilePath": "log4j-core-2.18.0.jar"
        }
      ]
    }
  ]
}

```

> You mean like in busybox? Results section doesn't appear at all in two cases: 1. OS not supported and there's no lang packages 2. There's no OS packages at all ?

We don't include OS packages in `Results` if OS is not supported or no OS packages are found. If other types (language files, secrets, misconfigs, etc.) are not found, the `Results` will be nil.

**Reply by Bah-bah** [reply]
Hi, @DmitriyLewen 
A small question: is it safe to assume that when no OS packages are found, there are no OS vulnerabilities in this image? Or they are just left out of Trivy's scope?

**Reply by DmitriyLewen** [reply]
When Trivy finds OS packages, Trivy checks for vulnerabilities for those packages (you must enable `vunl`scanner for  `cyclonedx` format).
That is, if the results table is empty, but the packages are detected, the OS packages do not contain vulnerabilities.
